File: swiftly/Services/SessionService.cs
using System;
using System.Collections.Generic;

namespace NotifyMessages;

/// Сервис сессий игроков: таймеры подключения и набор fully-connected.
///
/// Таймер хранится как «как его остановить», а не как тип фреймворка: класс одинаков
/// в cssharp/ и swiftly/, где таймеры устроены по-разному.
public sealed class SessionService
{
    private readonly object _lock = new();
    private readonly Dictionary<ulong, Action> _connectionTimers = new();
    private readonly HashSet<ulong> _fullyConnectedPlayers = new();

    // Язык клиента (cl_language), снятый при подключении. Чистится в EventPlayerDisconnect
    // вместе с остальным состоянием игрока — иначе словарь растёт всё время жизни сервера.
    private readonly Dictionary<ulong, string> _languages = new();

    // Игроки, вернувшиеся на сервер после смены карты, и момент возврата.
    // В CS2 при смене карты player_disconnect не приходит, а player_connect_full приходит
    // заново: второй connect_full у уже fully-connected игрока — это возврат, а не заход.
    // Его первое попадание в команду на новой карте — тоже не событие, и метка живёт
    // до этого момента (или до истечения срока, чтобы не съесть настоящий переход позже).
    private readonly Dictionary<ulong, DateTime> _returning = new();

    public void SetConnectionTimer(ulong steamId, Action stopTimer)
    {
        lock (_lock)
        {
            if (_connectionTimers.TryGetValue(steamId, out var existing))
                existing();
            _connectionTimers[steamId] = stopTimer;
        }
    }

    public bool TryKillAndRemoveConnectionTimer(ulong steamId)
    {
        lock (_lock)
        {
            if (_connectionTimers.TryGetValue(steamId, out var stop))
            {
                stop();
                _connectionTimers.Remove(steamId);
                return true;
            }
            return false;
        }
    }

    public void RemoveConnectionTimer(ulong steamId)
    {
        lock (_lock)
        {
            _connectionTimers.Remove(steamId);
        }
    }

    public void SetLanguage(ulong steamId, string? language)
    {
        if (string.IsNullOrWhiteSpace(language)) return;

        lock (_lock)
        {
            _languages[steamId] = language;
        }
    }

    public string? GetLanguage(ulong steamId)
    {
        lock (_lock)
        {
            return _languages.TryGetValue(steamId, out var lang) ? lang : null;
        }
    }

    public void RemoveLanguage(ulong steamId)
    {
        lock (_lock)
        {
            _languages.Remove(steamId);
        }
    }

    public void MarkReturning(ulong steamId, DateTime now)
    {
        lock (_lock)
        {
            _returning[steamId] = now;
        }
    }

    /// Снимает метку возврата и говорит, была ли она свежей. Метка снимается в любом
    /// случае: просроченная не должна ждать следующего перехода.
    public bool TakeReturning(ulong steamId, DateTime now, TimeSpan maxAge)
    {
        lock (_lock)
        {
            if (!_returning.Remove(steamId, out var markedAt)) return false;
            return now - markedAt <= maxAge;
        }
    }

    public void AddFullyConnected(ulong steamId)
    {
        lock (_lock)
        {
            _fullyConnectedPlayers.Add(steamId);
        }
    }

    public bool IsFullyConnected(ulong steamId)
    {
        lock (_lock)
        {
            return _fullyConnectedPlayers.Contains(steamId);
        }
    }

    public void RemoveFullyConnected(ulong steamId)
    {
        lock (_lock)
        {
            _fullyConnectedPlayers.Remove(steamId);
            _returning.Remove(steamId);
        }
    }

    public void Clear()
    {
        lock (_lock)
        {
            foreach (var kv in _connectionTimers) kv.Value();
            _connectionTimers.Clear();
            _fullyConnectedPlayers.Clear();
            _languages.Clear();
            _returning.Clear();
        }
    }
}

File: metamod/src/core/session_service.cpp
#include "core/session_service.h"

#include <utility>
#include <vector>

namespace nm {

void SessionService::SetConnectionTimer(uint64_t steamId, std::function<void()> stopTimer) {
    const auto existing = _connectionTimers.find(steamId);
    if (existing != _connectionTimers.end() && existing->second) existing->second();
    _connectionTimers[steamId] = std::move(stopTimer);
}

bool SessionService::TryKillAndRemoveConnectionTimer(uint64_t steamId) {
    const auto timer = _connectionTimers.find(steamId);
    if (timer == _connectionTimers.end()) return false;

    // Сначала убрать, потом звать: остановка не должна видеть себя в словаре
    std::function<void()> stop = std::move(timer->second);
    _connectionTimers.erase(timer);
    if (stop) stop();
    return true;
}

void SessionService::RemoveConnectionTimer(uint64_t steamId) { _connectionTimers.erase(steamId); }

void SessionService::SetLanguage(uint64_t steamId, const std::string& language) {
    bool blank = true;
    for (const char c : language) blank = blank && (c == ' ' || c == '\t' || c == '\r' || c == '\n');
    if (!blank) _languages[steamId] = language;
}

std::string SessionService::GetLanguage(uint64_t steamId) const {
    const auto language = _languages.find(steamId);
    return language != _languages.end() ? language->second : std::string();
}

void SessionService::RemoveLanguage(uint64_t steamId) { _languages.erase(steamId); }

void SessionService::MarkReturning(uint64_t steamId, double now) { _returning[steamId] = now; }

bool SessionService::TakeReturning(uint64_t steamId, double now, double maxAgeSeconds) {
    const auto mark = _returning.find(steamId);
    if (mark == _returning.end()) return false;

    const double markedAt = mark->second;
    _returning.erase(mark);
    return now - markedAt <= maxAgeSeconds;
}

void SessionService::AddFullyConnected(uint64_t steamId) { _fullyConnected.insert(steamId); }

bool SessionService::IsFullyConnected(uint64_t steamId) const { return _fullyConnected.count(steamId) != 0; }

void SessionService::RemoveFullyConnected(uint64_t steamId) {
    _fullyConnected.erase(steamId);
    _returning.erase(steamId);
}

void SessionService::Clear() {
    std::map<uint64_t, std::function<void()>> timers;
    timers.swap(_connectionTimers);
    for (auto& timer : timers) {
        if (timer.second) timer.second();
    }
    _fullyConnected.clear();
    _languages.clear();
    _returning.clear();
}

namespace team_swap {

constexpr int kTerrorists = 2;
constexpr int kCounterTerrorists = 3;

bool IsSideSwitch(int oldTeam, int newTeam) {
    return (oldTeam == kTerrorists && newTeam == kCounterTerrorists) ||
           (oldTeam == kCounterTerrorists && newTeam == kTerrorists);
}

bool IsMassSwap(const std::vector<std::pair<int, int>>& changes) {
    if (changes.size() < 2) return false;
    for (const auto& change : changes) {
        if (!IsSideSwitch(change.first, change.second)) return false;
    }
    return true;
}

}  // namespace team_swap
}  // namespace nm
